Treon test for demos

habiba-h opened the pull request:
### Description
`/scripts/test_demos.py` : Script to download the data and test demo within specified subdirectories of the demo folder with `Treon`. 
`/docker/stellargraph-treon/Dockerfile`: The Dockerfile for the demos is updated to run the above script for testing demos.  

Note, the script test  the demos in the specified subdirectories. In this way those demos that are not compatible for automated testing, such as, the dataset on which the notebook runs being unavailable, or requiring a myriad of libraries to be installed, can be excluded. 

Although the script runs smoothly as a standalone on the the demos, the docker image is not stable. It arbitrarily breaks down with `Kernel died` error resulting in failing at successfully testing the notebook it died on. This issue needs to be investigated. 
 
This functionality will enable testing of all demo notebooks in a CI in the future.

### Review Checklist (Development)
- [x] Write script to automate downloading of datasets & running of demo notebooks.
- [x] Branch and Pull Request build on CI
- [x] Branch and Pull Request pass unit tests on CI
- [ ] Version number reflects new status
- [x] Peer Code Review Performed
- [x] Code well commented
- [ ] Documentation in repo
- [ ] Team demo

adocherty commented:
I have updated some files to resolve errors I got when running the notebooks:

```
ERROR in testing ./calibration-pubmed-link-prediction.ipynb
            Traceback (most recent call last):

---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-45-2641839153f2> in <module>
----> 1 lr.fit(val_predictions, edge_labels_val)

~/Code/stellargraph/stellargraph/utils/calibration.py in fit(self, x_train, y_train)
    399             if len(x_train.shape) > 1:
    400                 x_train = x_train.reshape(-1)
--> 401             self.regressors[-1].fit(X=x_train, y=y_train)
    402         else:
    403             for n in range(self.n_classes):

~/.envs/stellargraph/lib/python3.6/site-packages/sklearn/isotonic.py in fit(self, X, y, sample_weight)
    333         # Transform y by running the isotonic regression algorithm and
    334         # transform X accordingly.
--> 335         X, y = self._build_y(X, y, sample_weight)
    336
    337         # It is necessary to store the non-redundant part of the training set

~/.envs/stellargraph/lib/python3.6/site-packages/sklearn/isotonic.py in _build_y(self, X, y, sample_weight, trim_duplicates)
    271         X, y, sample_weight = [array[order] for array in [X, y, sample_weight]]
    272         unique_X, unique_y, unique_sample_weight = _make_unique(
--> 273             X, y, sample_weight)
    274
    275         # Store _X_ and _y_ to maintain backward compat during the deprecation

sklearn/_isotonic.pyx in sklearn._isotonic._make_unique()

ValueError: Buffer dtype mismatch, expected 'float' but got 'double'
ValueError: Buffer dtype mismatch, expected 'float' but got 'double'
```

A similar error is encountered in the calibration notebook for link prediction:
```
ERROR in testing ./calibration-pubmed-link-prediction.ipynb

---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-45-2641839153f2> in <module>
----> 1 lr.fit(val_predictions, edge_labels_val)

~/.local/lib/python3.6/site-packages/stellargraph/utils/calibration.py in fit(self, x_train, y_train)
    399             if len(x_train.shape) > 1:
    400                 x_train = x_train.reshape(-1)
--> 401             self.regressors[-1].fit(X=x_train, y=y_train)
    402         else:
    403             for n in range(self.n_classes):

~/.local/lib/python3.6/site-packages/sklearn/isotonic.py in fit(self, X, y, sample_weight)
    333         # Transform y by running the isotonic regression algorithm and
    334         # transform X accordingly.
--> 335         X, y = self._build_y(X, y, sample_weight)
    336
    337         # It is necessary to store the non-redundant part of the training set

~/.local/lib/python3.6/site-packages/sklearn/isotonic.py in _build_y(self, X, y, sample_weight, trim_duplicates)
    271         X, y, sample_weight = [array[order] for array in [X, y, sample_weight]]
    272         unique_X, unique_y, unique_sample_weight = _make_unique(
--> 273             X, y, sample_weight)
    274
    275         # Store _X_ and _y_ to maintain backward compat during the deprecation

sklearn/_isotonic.pyx in sklearn._isotonic._make_unique()

ValueError: Buffer dtype mismatch, expected 'float' but got 'double'
ValueError: Buffer dtype mismatch, expected 'float' but got 'double'
```

These errors are most likely related to the version of scikit learn I am using (0.21.2). These have been fixed on the develop branch, so I have merged develop into this branch.

adocherty commented:
The docker issues with the Jupyter kernel dying could be related to this issue:
https://github.com/jupyter/jupyter_client/issues/154

I have run the test script using their suggestions and it seems to work. See the edited `Dockerfile`. I also increased the memory available to docker to 6GB to ensure there were no issues with crashes related to the demos running out of memory.

habiba-h commented:
@adocherty, tested the docker image. The script ran through without any issues.

adocherty commented:
FYI: we will need to wait for feedback from @TheDen before doing anything with this branch.

TheDen commented on docker/stellargraph-treon/Dockerfile:
```diff
@@ -0,0 +1,40 @@
+# keep pip up to date for the end user
+# hadolint ignore=DL3013
+RUN pip install --upgrade pip --user \
+      && pip install --no-cache-dir jupyter=="${JUPYTER_VERSION}" treon seaborn --user  
```
There are trailing spaces after `--user`

TheDen replied:
Rule [DL3013](https://github.com/hadolint/hadolint/wiki/DL3013) ensure versions of pip packages are pinned. It might be worthwhile to pin `treon` and `seaborn` versions as `jupyter` is, and remove the ignore.

habiba-h replied:
Pinned Treon adn Seaborn versions.

TheDen commented on docker/stellargraph-treon/Dockerfile:
```diff
@@ -0,0 +1,40 @@
+COPY --chown=stellar --from=base /data/Pubmed-Diabetes /home/stellar/data/pubmed/Pubmed-Diabetes
+COPY --chown=stellar --from=base /data/ml-100k /home/stellar/data/ml-100k
+COPY --chown=stellar --from=base /data/BlogCatalog-dataset /home/stellar/data/BlogCatalog-dataset
+COPY --chown=stellar ./scripts/ /home/stellar/scripts
```
Maybe move Line 38 to be below line 33, so `COPY`s from auxiliary images are after local COPYs

habiba-h replied:
Moved Line 38 below 33.

adocherty commented:
In discussions with @TheDen we should do the following before we merge this:
* Remove movielens downloads and demos. This dataset is not approved for distribution.
* Attempt to speed up the demos by limiting the number of training epochs, where we can. This practically means adding flags to the demo scripts (e.g. `-e 5` for script "gcn-cora-example.py").

After this we can look as using a dataset that is approved for distribution for the movielens example.

habiba-h commented:
@adocherty & @TheDen, 
I have incorporated the following changes you guys suggested: 
* Remove movielens downloads and demos. 
* Added the instructions to run the interpretability demo but left it commented at the moment until its updated based on the latest changes in the library.
* Added epochs  flags to the demo scripts. They run for 5 epochs each.
* Pinned Treon adn Seaborn versions and removed the trailing spaces after --users.
* Moved the line 38 below 33. 

Tested the docker image. It runs smoothly. However, I did get "Kernal Died" error once on the ensemble demo that takes the longest to run anyways. It failed only once though.

adocherty commented:
Looks good to me. Let's wait for @TheDen to check it then we can merge it.

TheDen commented on docker/stellargraph-treon/Dockerfile:
```diff
@@ -0,0 +1,43 @@
+      && tar --no-same-owner -xf cora.tgz \
+      && curl -sLO "https://linqs-data.soe.ucsc.edu/public/Pubmed-Diabetes.tgz" \
+      && tar --no-same-owner -xf Pubmed-Diabetes.tgz \
+     # && curl -sLO "http://files.grouplens.org/datasets/movielens/ml-100k.zip" \
```
We might as well delete these lines

habiba-h replied:
Deleted.

TheDen commented on docker/stellargraph-treon/Dockerfile:
```diff
@@ -0,0 +1,43 @@
+ENV JUPYTER_VERSION="1.0.0" \
+  PATH=${PATH}:/home/stellar/.local/bin
+  
+ENV TREON_VERSION="0.1.2" \
```
These can probably be merged in one `ENV`, e.g.,

```
ENV PATH=${PATH}:/home/stellar/.local/bin \
  JUPYTER_VERSION="1.0.0" \
  TREON_VERSION="0.1.2" \
  SEABORN_VERSION="0.9.0"
```

habiba-h replied:
Merged all the version instructions in one command.

TheDen commented on docker/stellargraph-treon/Dockerfile:
```diff
@@ -0,0 +1,43 @@
+COPY --chown=stellar ./scripts/ /home/stellar/scripts
+COPY --chown=stellar --from=base /data/cora /home/stellar/data/cora
+COPY --chown=stellar --from=base /data/Pubmed-Diabetes /home/stellar/data/pubmed/Pubmed-Diabetes
+#COPY --chown=stellar --from=base /data/ml-100k /home/stellar/data/ml-100k
```
This line can also be deleted

habiba-h replied:
Deleted.

adocherty commented:
This looks good now, let's merge it 👍
